src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -55,19 +79,126 @@ public async Task Invoke(HttpContext context)
             context.Response.StatusCode = _rejectionStatusCode;
-            await _onRejected(context, lease);
+
+            // If the global limiter rejected this request, use OnRejected from the options if it exists.
+            // Else the endpoint limiter rejected this request - use OnRejected from the endpoint policy if it exists,
+            // else OnRejected from the options if it exists.
+            if (leaseContext.Rejector.Equals(Rejector.Global))
+            {
+                if (_onRejected is not null)
+                {
+                    await _onRejected(new OnRejectedContext() { HttpContext = context, Lease = leaseContext.Lease }, context.RequestAborted);
+                }
+            }
+            else
+            {
+                DefaultRateLimiterPolicy? policy;
+                var name = context.GetEndpoint()?.Metadata.GetMetadata<IRateLimiterMetadata>()?.Name;
+                // Use custom policy OnRejected if available, else use OnRejected from the Options if available.
+                if (name is not null && _policyMap.TryGetValue(name, out policy) && policy.OnRejected is not null)
+                {
+                    await policy.OnRejected(new OnRejectedContext() { HttpContext = context, Lease = leaseContext.Lease }, context.RequestAborted);
+                }
+                else if (_onRejected is not null)
+                {
+                    await _onRejected(new OnRejectedContext() { HttpContext = context, Lease = leaseContext.Lease }, context.RequestAborted);
+                }
+            }
+        }
+    }
+
+    private ValueTask<LeaseContext> TryAcquireAsync(HttpContext context)
+    {
+        var leaseContext = CombinedAcquire(context);
+        if (leaseContext.Lease.IsAcquired)
+        {
+            return ValueTask.FromResult(leaseContext);
+        }
+
+        return CombinedWaitASync(context, context.RequestAborted);
+    }
+
+    private LeaseContext CombinedAcquire(HttpContext context)
+    {
+        RateLimitLease? globalLease = null;
+        RateLimitLease? endpointLease = null;
+
+        try
+        {
+            if (_globalLimiter is not null)
+            {
+                globalLease = _globalLimiter.Acquire(context);
+                if (!globalLease.IsAcquired)
+                {
+                    return new LeaseContext() { Rejector = Rejector.Global, Lease = globalLease };
+                }
+            }
+            endpointLease = _endpointLimiter.Acquire(context);
+            if (!endpointLease.IsAcquired)
+            {
+                globalLease?.Dispose();
+                return new LeaseContext() { Rejector = Rejector.Endpoint, Lease = endpointLease };
+            }
+        }
+        catch (Exception)
+        {
+            endpointLease?.Dispose();
+            globalLease?.Dispose();
+            throw;
         }
+
+        return new LeaseContext() { Lease = new DefaultCombinedLease(globalLease, endpointLease)};
```

Comment:
If globalLease is null then don't use DefaultCombinedLease. That way you can remove the null checks from DefaultCombinedLease.

Reply:
This class is essentially the same as https://github.com/dotnet/runtime/blob/6f7064dc170a46d1a5e67823f5178b8862b27a58/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ChainedPartitionedRateLimiter.cs#L159-L246 but with 2 args instead of an array - should we just make that public? CC @BrennanConroy

Reply:
That doesn't address the first point: If globalLease is null then you should directly return endpointLease without creating another object.